File: backend/python-capability-library/app/services/instruments/instrument_timbre_catalog.py
from __future__ import annotations

from copy import deepcopy
import json
from pathlib import Path
from typing import Any


_CATALOG_PATH = Path(__file__).resolve().parents[3] / "contracts" / "music" / "instrument-timbre-catalog.v1.json"
_CATALOG: dict[str, Any] = json.loads(_CATALOG_PATH.read_text(encoding="utf-8"))

INSTRUMENT_TIMBRE_CATALOG_VERSION = str(_CATALOG["version"])
SAMPLE_LIBRARY: dict[str, Any] = deepcopy(_CATALOG["sampleLibrary"])
EXACT_TIMBRE_DEFINITIONS: tuple[dict[str, Any], ...] = tuple(deepcopy(_CATALOG["exactTimbres"]))
PENDING_EXACT_TIMBRE_DEFINITIONS: tuple[dict[str, Any], ...] = tuple(
    deepcopy(_CATALOG["pendingExactTimbres"])
)


def list_exact_timbres() -> list[dict[str, Any]]:
    return deepcopy(list(EXACT_TIMBRE_DEFINITIONS))


def list_pending_exact_timbres() -> list[dict[str, Any]]:
    return deepcopy(list(PENDING_EXACT_TIMBRE_DEFINITIONS))

File: frontend/review-console/src/shared/instrumentTimbreCatalog.ts
import catalogData from "../../../../contracts/music/instrument-timbre-catalog.v1.json";

export type TimbrePreview = {
  baseMidi: number;
  offsets: number[];
  gap: number;
  duration: number;
  gain: number;
};

export type ExactTimbreDefinition = {
  id: string;
  label: string;
  familyId: "keyboard" | "plucked_string" | "bowed_string" | "woodwind" | "brass" | "free_reed" | "percussion";
  familyLabel: string;
  classroomFamily: string;
  playbackInstrument: string;
  sampleFidelity: "exact_open_sample";
  evidenceTerms: string[];
  classroomNote: string;
  preview: TimbrePreview;
};

export type PendingExactTimbreDefinition = {
  id: string;
  label: string;
  familyLabel: string;
  playbackInstrument: string;
  sampleFidelity: "close_soundfont_sample" | "approximate_soundfont_sample";
  evidenceTerms: string[];
  classroomNote: string;
  requiredExactAsset: string;
  preview: TimbrePreview;
};

export type TimbreSampleLibrary = {
  id: string;
  format: string;
  license: string;
  licenseUrl: string;
  sourceUrl: string;
  attribution: string;
};

type InstrumentTimbreCatalog = {
  version: string;
  sampleLibrary: TimbreSampleLibrary;
  exactTimbres: ExactTimbreDefinition[];
  pendingExactTimbres: PendingExactTimbreDefinition[];
};

const catalog = catalogData as InstrumentTimbreCatalog;

export const INSTRUMENT_TIMBRE_CATALOG_VERSION = catalog.version;
export const SAMPLE_LIBRARY = catalog.sampleLibrary;
export const EXACT_TIMBRE_DEFINITIONS = catalog.exactTimbres;
export const PENDING_EXACT_TIMBRE_DEFINITIONS = catalog.pendingExactTimbres;
